File: challenges/0003-burrows-wheeler/solution.py
from sys import stdin


class Shift:
    def __init__(self, index, s):
        self.index = index
        self.string = s
        self.rank = (-1, -1)

    def __lt__(self, other):
        if not isinstance(other, Shift):
            raise Exception('Invalid argument for comparison with Shift')
        return self.rank < other.rank


def sorted_cyclic_shifts(s):
    """Sort all cyclic shifts of a string.

    Code from http://cp-algorithms.com/string/suffix-array.html.
    """
    n = len(s)
    shifts = [Shift(j, s) for j in range(n)]
    ranks = [ord(c) for c in s]
    length, limit = 1, 2 * n
    while length < limit:
        h = length // 2
        for shift in shifts:
            x = ranks[shift.index]
            y = ranks[(shift.index + h) % n]
            shift.rank = (x, y)
        shifts.sort()

        rank = 0
        ranks[shifts[0].index] = rank
        for j in range(1, n):
            previous, current = shifts[j-1], shifts[j]
            if previous.rank != current.rank:
                rank += 1
            ranks[current.index] = rank

        length *= 2

    return [shift.index for shift in shifts]


def bwt(s):
    """Burrows-Wheeler transform."""
    shifts = sorted_cyclic_shifts(s)
    t = "".join(s[j-1] for j in shifts)
    return t, shifts.index(0)


def inverse_bwt(t, k):
    """Inverse Burrows-Wheeler transform."""
    n = len(t)
    f = sorted((ch, j) for j, ch in enumerate(t))
    chars = []
    for _ in range(n):
        ch, j = f[k]
        chars.append(ch)
        k = j
    return "".join(chars)


print('\n'.join(map(lambda line: bwt(line[:-1])[0], stdin)))

File: challenges/0003-burrows-wheeler/solution.cxx
#include <iostream>
#include <string>
#include <vector>
#include <algorithm>
#include <cctype>
#include <tuple>
#include <iterator>

using namespace std;

typedef struct shift {
    int index, first_half_rank, second_half_rank;
    shift(): index(-1), first_half_rank(-1), second_half_rank(-1) {}
} suffix;

bool operator <(const shift &s, const shift &o)
{
   return make_tuple(s.first_half_rank, s.second_half_rank) < make_tuple(o.first_half_rank, o.second_half_rank);
}


vector<shift> sorted_cyclic_shifts(const string &s)
{
    int n = static_cast<int>(size(s));
    vector<shift> shifts(n);
    vector<int> ranks(n);

    for (int i = 0; i < n; ++i) {
        shifts[i].index = i;
        ranks[i] = s[i] - 32;
    }

    int limit = 2 * n;

    for (int length = 2; length < limit; length *= 2) {
        auto h = length / 2;

        for (int i = 0; i < n; ++i) {
            shifts[i].first_half_rank = ranks[shifts[i].index];
            shifts[i].second_half_rank = ranks[(shifts[i].index + h) % n];
        }

        sort(begin(shifts), end(shifts));

        int current_rank = 0;

        ranks[shifts[0].index] = current_rank;
        for (int i = 1, current_rank = 0; i < n; ++i) {
            auto previous_shift_rank = make_tuple(shifts[i-1].first_half_rank, shifts[i-1].second_half_rank);
            auto current_shift_rank = make_tuple(shifts[i].first_half_rank, shifts[i].second_half_rank);

            if (previous_shift_rank != current_shift_rank) {
                ++current_rank;
            }

            ranks[shifts[i].index] = current_rank;
        }
    }

    return shifts;
}

int main()
{
    string s, t;

    while (getline(cin, s)) {
        int n = static_cast<int>(size(s));
        auto shifts = sorted_cyclic_shifts(s);
        string t;

        for (const auto &sh: shifts) {
            t.push_back(s[(sh.index + n - 1) % n]);
        }

        cout << t << endl;
    }

    return 0;
}

